enhancements/ingress/global-admission-hsts.md

```diff
@@ -0,0 +1,631 @@
+for reasons provided within this proposal.  Thanks to David Eads and Miciah Masters for their insights and contributions to the updated proposal.
+
+## Motivation
+
+HSTS ([RFC 6797](https://tools.ietf.org/html/rfc6797)) policy enforces the use
+of HTTPS in client requests to the host, without having to make use of HTTP redirects.
+HSTS provides user protection and is concerned with minimizing security threats based on
+network traffic eavesdropping and man-in-the-middle attacks.  Using a response header
+called `Strict-Transport-Security`, the HTTP response informs clients that a website
+can only be accessed via HTTPS.
+
+Administrators who are tasked with route management and/or regulatory compliance face a
+number of issues with regard to enforcing HSTS.  For efficiency and protection against
+configuration errors, they have requested to automatically and globally enable HSTS
+on the basis of the cluster `Ingress` domains. However, because enabling HSTS automatically
+and globally can cause disruption of service on a wide scale, they should also be able to
+audit and change HSTS configuration without further outages.  Finally, they should be able
+to consistently apply the same HSTS configuration and predict the outcome on any cluster.
+
+### Goals
+
+This proposal allows administrators to:
+- Continue to enable HSTS per-route (retain the existing feature functionality)
+- Provide HSTS verification per-domain
+- Audit HSTS configurations on any cluster
+- Predict HSTS enforcement on any cluster with a single configuration manifest
+
+### Non-Goals
+
+- HSTS cannot be applied to non-TLS routes, even if HSTS is requested for all routes globally.  As
+mentioned in [RFC 6797 Section 8.1](https://tools.ietf.org/html/rfc6797#section-8.1), if an HTTP
+response is received over insecure transport, clients must ignore any present STS headers.
+
+## Proposal
+Each goal is addressed in the subsections below.
+
+### Continue to enable HSTS per-route (retain the existing feature functionality)
+The `Strict-Transport-Security` response header is currently applied per-route via an
+annotation called `haproxy.router.openshift.io/hsts_header`.  This annotation has a
+required directive `max-age`, and optional directives `includeSubDomains` and `preload`,
+with the following definitions:
+- `max-age`: delta time range in seconds during which the host is to be regarded as an HSTS host
+  - If set to 0, it negates the effect and the host is no longer regarded as an HSTS host
+  - `max-age` is a time-to-live value, so if a client makes a request for the route's host
+  and this period of time elapses before the client makes another request for the same host, the
+  HSTS policy will expire on that client
+- `includeSubDomains`: if present, applies to any subdomains of the host's domain name, when the
+route admission policy allows wildcards.  If wildcard routes aren't allowed, this directive has no
```

Comment:
> If wildcard routes aren't allowed, this directive has no effect

Clarify this for me.  If foo.company.com returns an HSTS header that says includeSubDomains, even if our router doesn't serve, superimportant-http-resource.foo.company.com will suddenly become inaccessible, right?  That's part of the risk we're looking to mitigate.

Reply:
Thanks!  I updated this section, code samples, and the risks and mitigation sections.

Reply:
If the route admission policy with host `foo.com` allows subdomain wildcards it should use the `includeSubDomains` directive only if all subdomains of that route should share the same HSTS Policy.  There cannot be a mix of plain HTTP and HTTPS routes covered by the route. 

If the route admission policy with host `foo.com` does not allow subdomain wildcards, it may not use the `includeSubDomains` directive, because it would unintentionally convey its HSTS Policy to any routes having `foo.com` as a superdomain.  A route configured without allowing subdomain wildcards will be rejected if it contains `includeSubdomains` in its HSTS Policy.

Reply:
And to be clear, while we only have visibility to route based ingress into a single cluster, there is no actual restriction that our router owns an entire subdomain.  Traffic can be intercepted in front of it using ServerName from the TCP requests themselves to have a subdomain of foo.com served by something that isn't even openshift based.

Reply:
In the end, there will be no changes in whether a route is accepted/rejected.  I will document all the ways things could go wrong, for the end user, and let them do what they choose.  I listed this in Design Details in what I hope is the last update of this EP, plus added a few corrections.